File: TPII-2025-2/PadraoGoF/Criacional/Builder/ExercicioPizzaria.py
#Definindo partes da pizza
class Tamanho:
    def __init__(self, tipoTamanho):
        self.tipoTamanho = tipoTamanho

class Massa:
    def __init__(self, tipoMassa):
        self.tipoMassa = tipoMassa

class Molho:
    def __init__(self, tipoMolho):
        self.tipoMolho = tipoMolho

class Cobertura:
    def __init__(self, tipoCobertura):
        self.tipoCobertura = tipoCobertura

#Builder
class PizzaBuilder:
    def __init__(self):
        self.tamanho
        self.massa
        self.molho
        self.cobertura

    def addTamanho(self, tipoTamanho):
        self.tamanho = Tamanho(tipoTamanho)
        return self
    
    def addMassa(self, tipoMassa):
        self.massa = Massa(tipoMassa)
        return self
    
    def addMolho(self, tipoMolho):
        self.molho = Molho(tipoMolho)
        return self
    
    def addCobertura(self, tipoCobertura):
        self.cobertura = Cobertura(tipoCobertura)
        return self
    
    def construir(self):
        pizza = Pizza(self.tamanho, self.massa, self.molho, self.cobertura)
        return pizza
    
#Construir pizza
class Pizza:
    def __init__(self, tamanho, massa, molho, cobertura):
        self.tamanho = tamanho
        self.massa = massa
        self.molho = molho
        self.cobertura = cobertura

    def mostrarDetalhes(self):
        print(f"Pizza com tamanho {self.tamanho.tipoTamanho}\n massa: {self.massa.tipoMassa}\n molho: {self.molho.tipoMolho}\n cobertura: {self.cobertura.tipoCobertura}\n")

#Uso do modelo builder - construindo uma pizza
pizzaDePortuguesa = (
    PizzaBuilder()
    .addTamanho('Grande')
    .addMassa('Fina')
    .addMolho('Vermelho')
    .addCobertura('Queijo')
    .construir()
)

File: TPII-2025-2/PadraoGoF/Criacional/Builder/ExercicioPizzaria.js
//Definindo partes da pizza
class Tamanho{
    constructor(tipoTamanho){
        this.tipoTamanho = tipoTamanho;
    }
}

class Massa{
    constructor(tipoMassa){
        this.tipoMassa = tipoMassa;
    }
}

class Molho{
    constructor(tipoMolho){
        this.tipoMolho = tipoMolho;
    }
}

class Cobertura{
    constructor(tipoCobertura){
        this.tipoCobertura = tipoCobertura;
    }
}

//Builder
class PizzaBuilder{ 
    constructor(){
    this.reset();
    }

    reset(){
        this.tamanho = null;
        this.massa = null;
        this.molho = null;
        this.cobertura = [];
    }

    addTamanho(tipoTamanho){
        this.tamanho = new Tamanho(tipoTamanho);
        return this;
    }

    addMassa(tipoMassa){
        this.massa = new Massa(tipoMassa);
        return this;
    }

    addMolho(tipoMolho){
        this.molho = new Molho(tipoMolho);
        return this;
    }

    addCobertura(tipoCobertura){
        this.cobertura.push(tipoCobertura);
        return this;
    }

    construir(){
        return new Pizza(this.tamanho, this.massa, this.molho, this.cobertura);
    }
}

//Construir pizza
class Pizza{
    constructor(tamanho, massa, molho, cobertura){
        this.tamanho = tamanho;
        this.massa = massa;
        this.molho = molho;
        this.cobertura = cobertura;
    }

    mostrarDetalhes(){
        console.log(`Pizza com tamanho ${this.tamanho.tipoTamanho}\n
             massa: ${this.massa.tipoMassa}\n 
             molho: ${this.molho.tipoMolho}\n 
             cobertura: ${this.cobertura.tipoCobertura}\n`
            )
    }
}

//Uso do modelo builder - construindo uma pizza
const builder = new PizzaBuilder();

const pizzaDePortuguesa = builder
    .addTamanho('Grande')
    .addMassa('Fina')
    .addMolho('Vermelho')
    .addCobertura('Queijo')
    .addCobertura('Tomate')
    .addCobertura('Orégano')
    .construir()

const pizzaDeChocolate = builder
    .addTamanho('Broto')
    .addMassa('Fina')
    .addMolho('Chocolate')
    .addCobertura('M&Ms')
    .addCobertura('Chocolate')
    .construir()

pizzaDePortuguesa.mostrarDetalhes();
pizzaDeChocolate.mostrarDetalhes();